File: IDAUtil/Model/Properties/TcodeProperty/VA02Obj/OrderObjects/Switches/SwitchesOrderPropertyFactory.vb
Public Class SwitchesOrderPropertyFactory

    Public Sub New()
    End Sub

    Public Function getSapSwitchObjectList(spl As List(Of SwitchesProperty)) As List(Of SwitchesSapOrderProperty)
        Dim sapSwitchObjectQuery As List(Of SwitchesSapOrderProperty) =
            (From q In spl
             Group q By line = New With {Key q.order, q.salesOrg, q.soldTo}
                                    Into Group
             Select New SwitchesSapOrderProperty With {
            .salesOrg = line.salesOrg,
            .order = line.order,
            .soldTo = line.soldTo,
            .lineDetails = Group.Aggregate(New List(Of SwitchesSapLineProperty),
                Function(x, q)
                    x.Add(New SwitchesSapLineProperty() With {
                        .lineNumber = q.item,
                        .oldSku = q.oldSku,
                        .newSku = q.newSku,
                        .isSameBarcode = (q.oldSkuCaseBarcode = q.newSkuCaseBarcode AndAlso q.oldSkuUnitBarcode = q.newSkuUnitBarcode),
                        .reason = $"{If(q.switchForCustomer = "All", "Country level switch. ", "Speciffic switch for " & q.soldTo & " - " & q.shipToName & ". ")} {If(q.needOutOfStockToSwitch = True, "Sku " & q.oldSku & " is out of stock for this order. (order qty - " & q.orderedQty & " confirmed qty - " & q.confirmedQty & " ) ", "Switch was performed regardless of stock levels. " & q.switchComment)}"})
                    Return x
                End Function)}).ToList

        Return sapSwitchObjectQuery
    End Function
End Class

File: IDAUtil/Model/Properties/TcodeProperty/VA02Obj/OrderObjects/Switches/SwitchesOrderPropertyFactory.cs
using IDAUtil.Model.Properties.TaskProperty;
using System.Collections.Generic;
using System.Data;
using System.Linq;

namespace IDAUtil {
    public class SwitchesOrderPropertyFactory {
        public SwitchesOrderPropertyFactory() {
        }

        public List<SwitchesSapOrderProperty> getSapSwitchObjectList(List<SwitchesProperty> spl) {
            var sapSwitchObjectQuery = (from q in spl
                                        group q by new { q.order, q.salesOrg, q.soldTo } into line
                                        select new SwitchesSapOrderProperty() {
                                            salesOrg = line.Key.salesOrg,
                                            orderNumber = line.Key.order,
                                            soldTo = line.Key.soldTo,
                                            lineDetails = line.Aggregate(new List<SwitchesSapLineProperty>(), (x, q) => {
                                                x.Add(new SwitchesSapLineProperty() {
                                                    lineNumber = q.item,
                                                    oldSku = q.oldSku,
                                                    newSku = q.newSku,
                                                    isSameBarcode = (q.oldSkuCaseBarcode ?? "") == (q.newSkuCaseBarcode ?? "") && (q.oldSkuUnitBarcode ?? "") == (q.newSkuUnitBarcode ?? ""),
                                                    reason = $"{((q.switchForCustomer ?? "") == "All" ? "Country level switch. " : "Speciffic switch for " + q.soldTo + " - " + q.shipToName + ". ")} {(q.needOutOfStockToSwitch == true ? "Sku " + q.oldSku + " is out of stock for this order. (order qty - " + q.orderedQty + " confirmed qty - " + q.confirmedQty + " ) " : "Switch was performed regardless of stock levels. " + q.switchComment)}"
                                                });
                                                return x;
                                            })
                                        }).ToList();
            return sapSwitchObjectQuery;
        }
    }
}